llms/mlx_lm/lora.py

```diff
@@ -0,0 +1,245 @@
+    )
+    parser.add_argument(
+        "--learning-rate", type=float, default=1e-5, help="Adam learning rate."
+    )
+    parser.add_argument(
+        "--steps-per-report",
+        type=int,
+        default=10,
+        help="Number of training steps between loss reporting.",
+    )
+    parser.add_argument(
+        "--steps-per-eval",
+        type=int,
+        default=200,
+        help="Number of training steps between validations.",
+    )
+    parser.add_argument(
+        "--resume-adapter-file",
+        type=str,
+        default=None,
+        help="Load path to resume training with the given adapter weights.",
+    )
+    parser.add_argument(
+        "--adapter-file",
+        type=str,
+        default="adapters.npz",
+        help="Save/load path for the trained adapter weights.",
+    )
+    parser.add_argument(
+        "--save-every",
+        type=int,
+        default=100,
+        help="Save the model every N iterations.",
+    )
+    parser.add_argument(
+        "--test",
+        action="store_true",
+        help="Evaluate on the test set after training",
+    )
+    parser.add_argument(
+        "--test-batches",
+        type=int,
+        default=500,
+        help="Number of test set batches, -1 uses the entire test set.",
+    )
+    parser.add_argument("--seed", type=int, default=0, help="The PRNG seed")
+    return parser
+
+
+class Dataset:
+    """
+    Light-weight wrapper to hold lines from a jsonl file
+    """
+
+    def __init__(self, path: Path, key: str = "text"):
+        if not path.exists():
+            self._data = None
+        else:
+            with open(path, "r") as fid:
+                self._data = [json.loads(l) for l in fid]
+        self._key = key
+
+    def __getitem__(self, idx: int):
+        return self._data[idx][self._key]
+
+    def __len__(self):
+        return len(self._data)
+
+
+def load_dataset(args):
+    names = ("train", "valid", "test")
+    train, valid, test = (Dataset(Path(args.data) / f"{n}.jsonl") for n in names)
+    if args.train and len(train) == 0:
+        raise ValueError(
+            "Training set not found or empty. Must provide training set for fine-tuning."
+        )
+    if args.train and len(valid) == 0:
+        raise ValueError(
+            "Validation set not found or empty. Must provide validation set for fine-tuning."
+        )
+    if args.test and len(test) == 0:
+        raise ValueError(
+            "Test set not found or empty. Must provide test set for evaluation."
+        )
+    return train, valid, test
+
+
+if __name__ == "__main__":
+    parser = build_parser()
+    args = parser.parse_args()
+
+    np.random.seed(args.seed)
+
+    print("Loading pretrained model")
+    model, tokenizer = load(args.model)
+
+    # Freeze all layers other than LORA linears
+    model.freeze()
+    for l in model.model.layers[len(model.model.layers) - args.lora_layers :]:
```

Comment:
Ideally, when applying Lora linear, it should be moved into the trainer. The user only needs to specify the number of Lora layers and provide a targeted linear layer name list. The trainer should then apply Lora linear based on those specifications. However, I couldn't figure out a good way to do that. Maybe @awni you have some thoughts on how to do it?

Reply:
Have we considered adopting an option for a yaml configuration file, such as #235 was working on? I think there can be a way to cleanly grab from a config file such as this, provide defaults in the case of missing params, but also allow users to pass their own layers without a yaml config file, for example in the case of rapid prototyping on a notebook? Also, we can probably have layer checking in a separate function somewhere. That would keep the logic somewhat clean. Let me know what you think

Reply:
Yeah, reading from a YAML config file is definitely a good option. Once this PR is merged, we can start exploring the options and determine the best way to proceed. Ideally, I think it would be great if we could provide a similar API like peft lib has in loraconfig which takes target_modules as input and automatically applies the LoRa based on that.
https://huggingface.co/docs/peft/tutorial/peft_model_config

Reply:
- I think it's a good idea to do it in steps (this PR followed by refactor improvements
- I agree this part of the code belongs in the tuner. 
- I would say a high level spec in a json or yaml (I happen to prefer json 😄 ) is a good idea.